boltstub/packstream.py

```diff
@@ -44,22 +46,129 @@
 
 class Structure:
 
-    def __init__(self, tag, *fields):
+    def __init__(self, tag, *fields, verified=True):
         self.tag = tag
         self.fields = list(fields)
+        self._verified = verified
+        if verified:
+            self._verify_fields()
+
+    @property
+    def verified(self):
+        return self._verified
+
+    def _verify_fields(self):
+        tag, fields = self.tag, self.fields
+
+        def verify_node():
+            if (len(fields) != 3
+                    or not isinstance(fields[0], int)
+                    or not isinstance(fields[1], list)
+                    or not all(isinstance(label, str) for label in fields[1])
+                    or not isinstance(fields[2], dict)
+                    or not all(isinstance(k, str) for k in fields[2].keys())):
+                raise ValueError("Invalid Node struct received %r" % self)
+
+        def verify_relationship():
+            if (len(fields) != 5
+                    or not isinstance(fields[0], int)
+                    or not isinstance(fields[1], int)
+                    or not isinstance(fields[2], int)
+                    or not isinstance(fields[3], str)
+                    or not isinstance(fields[4], dict)
+                    or not all(isinstance(k, str) for k in fields[4].keys())):
+                raise ValueError(
+                    "Invalid Relationship struct received %r" % self
+                )
+
+        def verify_unbound_relationship():
+            if (len(fields) != 3
+                    or not isinstance(fields[0], int)
+                    or not isinstance(fields[1], str)
+                    or not isinstance(fields[2], dict)
+                    or not all(isinstance(k, str) for k in fields[2].keys())):
+                raise ValueError(
+                    "Invalid UnboundRelationship struct received %r" % self
+                )
+
+        def verify_path():
+            if (len(fields) != 3
+                    or not isinstance(fields[0], list)
+                    or not all(isinstance(n, Structure)
+                               and n.tag == b"\x4E"
```

Comment:
Maybe we could have this magic binaries in some "constant", it could be a bit difficult to keep track of this values sometimes.